### ISSUE TITLE
Flow: View can be a string?

### ISSUE THREAD

**Issue by Ashoat** [facebook/react-native#15955]
### Is this a bug report?

Yes

### Have you read the [Contributing Guidelines](https://facebook.github.io/react-native/docs/contributing.html)?

Yes

### Environment

1. `react-native -v`:
```
react-native-cli: 2.0.1
react-native: 0.48.2
```
2. `node -v`:
```
v8.4.0
```
3. `npm -v`:
```
5.4.1
```
4. `flow version`:
```
Flow, a static type checker for JavaScript, version 0.51.1
```

Then, specify:

- Target Platform: N/A, Flow static typesystem

- Development Operating System: macOS

- Build tools: N/A, Flow static typesystem

### Steps to Reproduce

1.
```js
render() {
  return <View ref={(view: ?View) => {}} />;
}
```
2. Flow errors out because it thinks View is a string.

### Expected Behavior

View should be a `React.Component`.

### Actual Behavior

1. [`View` is the result of calling `requireNativeComponent()`](https://github.com/facebook/react-native/blob/master/Libraries/Components/View/View.js#L117)
2. [`requireNativeComponent` returns `React$ComponentType<any> | string`](https://github.com/facebook/react-native/blob/master/Libraries/ReactNative/requireNativeComponent.js#L52)
3. Consequently, Flow thinks `View` can be a `string`

### Reproducible Demo

```js
// @flow

import React from 'react';
import { View } from 'react-native';

class Test extends React.PureComponent {

  render() {
    return <View ref={(view: ?View) => {}} />;
  }

}
```

**Comment by Ashoat** [comment]
@bvaughn - any chance you can elucidate why View can be a string? You added a type suppression relating to this [here](https://github.com/facebook/react-native/blob/master/Libraries/Components/ScrollView/ScrollView.js#L915). Looks like it has to do with something in React Fiber?

Can we not reliably depend on the ref returned from `View` being a `React.Component`? Or is the export from the `react-native` package typed incorrectly?

**Comment by bvaughn** [comment]
> Can we not reliably depend on the ref returned from `View` being a `React.Component`?

Unfortunately, that is correct. Native view components ("host components") are represented as strings with React 16. This is true for React DOM (eg "div", "span", etc.) and React Native (eg "RCTView", "RCTText", etc). `React.Component` (or `React.PureComponent`) are classes you extend to create JavaScript components.

Previously, React 15 created a wrapper object around host components (which is why the Flow typing "worked" before) but this was inefficient so we removed it for 16.

Internally, it looks like people are setting the ref-type to a mix of `any` or `NativeMethodsMixinType` (but I don't think this latter type is accessible to OSS users).

**Comment by Ashoat** [comment]
Thanks for responding!!

Any idea if there is a way to get to `View.measure()` in this new world? Unfortunately this old API is the only current way to fetch the absolute position of a `View` on the screen (`onLayout` doesn't have this info: #10556). Fetching the absolute position might seem unnecessary, but unfortunately `PanResponder`'s local position information is double broken (#12591, #15290), so absolute position information is necessary for the comparison.

Ideally the local position issues will be fixed, and if not hopefully the absolute positions will be added to the `onLayout` callback. But failing either of those things happening, `View.measure()` is currently the only lifeline for tracking the x, y coordinates of a pan relative to a View.

**Comment by Ashoat** [comment]
I had a new thought in bed last night: if people within Facebook are typing the parameter passed in to the `ref` callback prop as a `NativeMethodsMixinType`, that means it can't be a string correct?

It seems like the `T` in `React.Element<T>` - ie. the "tag name" used when writing JSX - can now be a string as well as a `React.Component`. But perhaps the parameter passed to the `ref` callback will continue being a `React.Component`, or at least something with `NativeMethodsMixinType`.

If that's the case, I just need to type the parameter to the `ref` callback differently (probably mirroring `NativeMethodsMixinType` (which I think I can sneakily include using an exact path into `react-native`).

My main question is: can the parameter to the `ref` callback be a string too? Or is that only the `View` imported from the `react-native` package?

**Comment by bvaughn** [comment]
That's correct @Ashoat, the value of `ref` is not a string. The value of `View` is a string though. So when you render a native `View` you're really creating an element with type "RCTView" (similar to how rendering an `HTMLDivElement` in the DOM is accomplished by creating an element with type "div"). The string just signals to React fiber which element type to create, native or not.

For native components, React Native renderer instantiates an instance of [`ReactNativeFiberHostComponent`](https://github.com/facebook/react/blob/master/src/renderers/native/ReactNativeFiberHostComponent.js) which is what gets passed to the ref callback. Unfortunately this type isn't exposed via the [React Native public API](https://github.com/facebook/react-native/blob/master/Libraries/react-native/react-native-implementation.js) so you can't use _it_ for the type yet.

For the time being, you could copy the following Flow types [from the react codebase](https://github.com/facebook/react/blob/master/src/renderers/native/ReactNativeTypes.js#L62-L73):
```js
type MeasureOnSuccessCallback = (
  x: number,
  y: number,
  width: number,
  height: number,
  pageX: number,
  pageY: number,
) => void;

type MeasureInWindowOnSuccessCallback = (
  x: number,
  y: number,
  width: number,
  height: number,
) => void;

type MeasureLayoutOnSuccessCallback = (
  left: number,
  top: number,
  width: number,
  height: number,
) => void;

// This would be the "ref" type for host components
type NativeMethodsMixinType = {
  blur(): void,
  focus(): void,
  measure(callback: MeasureOnSuccessCallback): void,
  measureInWindow(callback: MeasureInWindowOnSuccessCallback): void,
  measureLayout(
    relativeToNativeNode: number,
    onSuccess: MeasureLayoutOnSuccessCallback,
    onFail: () => void,
  ): void,
  setNativeProps(nativeProps: Object): void,
};
```

cc @calebmer: Let's talk about a way to make this work better for open source React Native + Flow users? (Maybe a way exists already that I'm unfamiliar with?)

**Comment by bvaughn** [comment]
Unfortunately I can't assign this issue to myself for some reason, but I'll create a PR shortly that will hopefully provide a meaningful default Flow type for host components. Will post back here once it's done.

**Comment by Ashoat** [comment]
My current solution is just:
```js
import type {
  NativeMethodsMixinType,
} from 'react-native/Libraries/Renderer/shims/ReactNativeTypes';
```
And then use `NativeMethodsMixinType` where I previously was using `View`.

**Comment by bvaughn** [comment]
Using `NativeMethodsMixinType` for your refs for now seems like the best solution, however you declare or import it. I'll try to land some .js.flow files for `View` and the other host component types so this will _just work_ in a future release and won't require any funky typing like this.

**Comment by bvaughn** [comment]
FYI this issue should be fixed by 11b4084. Someone (who has permissions to do so) can probably close this issue.
